jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java

```diff
@@ -64,33 +69,40 @@ public static FileEntriesLayer extraDirectoryLayerConfiguration(
           FileSystems.getDefault().getPathMatcher("glob:" + entry.getKey()), entry.getValue());
     }
 
-    new DirectoryWalker(sourceDirectory)
-        .filterRoot()
-        .walk(
-            localPath -> {
-              AbsoluteUnixPath pathOnContainer =
-                  targetDirectory.resolve(sourceDirectory.relativize(localPath));
-              Instant modificationTime = modificationTimeProvider.get(localPath, pathOnContainer);
-              FilePermissions permissions =
-                  extraDirectoryPermissions.get(pathOnContainer.toString());
-              if (permissions == null) {
-                // Check for matching globs
-                Path containerPath = Paths.get(pathOnContainer.toString());
-                for (Map.Entry<PathMatcher, FilePermissions> entry : pathMatchers.entrySet()) {
-                  if (entry.getKey().matches(containerPath)) {
-                    builder.addEntry(
-                        localPath, pathOnContainer, entry.getValue(), modificationTime);
-                    return;
-                  }
-                }
-
-                // Add with default permissions
-                builder.addEntry(localPath, pathOnContainer, modificationTime);
-              } else {
-                // Add with explicit permissions
-                builder.addEntry(localPath, pathOnContainer, permissions, modificationTime);
+    DirectoryWalker walker = new DirectoryWalker(sourceDirectory).filterRoot();
+    excludes
+        .stream()
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .forEach(pathMatcher -> walker.filter(path -> !pathMatcher.matches(path)));
+    includes
+        .stream()
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .map(pathMatcher -> (Predicate<Path>) (path -> pathMatcher.matches(path)))
+        .reduce((matches1, matches2) -> matches1.or(matches2))
+        .ifPresent(walker::filter);
+    walker.walk(
+        localPath -> {
```

Comment:
Hm this method is starting to get a little long, can we have line 83 - 105 in a separate method?

Reply:
The closure makes use of a lot of external variables, and mechanically refactoring out the closure results in a long list of arguments.
```java
  private static void addLayerEntry(
      Path sourceDirectory,
      AbsoluteUnixPath targetDirectory,
      Map<String, FilePermissions> extraDirectoryPermissions,
      ModificationTimeProvider modificationTimeProvider,
      Map<PathMatcher, FilePermissions> permissionPathMatchers,
      FileEntriesLayer.Builder layerBuilder,
      Path localPath) {
```
I realized most of the complication is due to determining permissions and calling `addEntry` based on that. So I refactored out the permissions part.

I also added a few more code comments in this method.

Reply:
Looks good. Putting the permissions logic in a helper helped a lot